### DISCUSSION TITLE
Code pour variabel

### DISCUSSION THREAD

**Discussion by Istomine** [web-classroom/web-classroom.github.io#49] [Q&A : Théorie / Cours] [answered]
```javascript
z = 1;

function x() {
    y = 1;
    var z = 2;
}

x();

console.log(y); 
console.log(z);
```

**Comment by Nortalle** [comment, marked as answer]
Bonsoir,

tldr : une variable est globale si elle est déclarée sans mot-clé et elle peut être redéclarée (JavaScript 🤯)

Pour le contexte : @Istomine est venu me voir en cours pour me poser une question à propos de ce code. À savoir qu'il y a cet output :

```js
z = 1;

function x() {
  y = 1;
  var z = 2;
}

x();

console.log(y); // 1
console.log(z); // 1
```

Tout d'abord. Nous pouvons voir dans cet exemple que nous pouvons déclarer une variable sans les mots-clés `var`, `let` ou `const`. La variable sera alors globale. [tutorialsteacher](https://www.tutorialsteacher.com/javascript/javascript-variable) · [w3schools](https://www.w3schools.com/js/js_variables.asp)

La question était de savoir quel était vraiment le scope de `var`.

> Les déclarations de variables sont traitées avant que le code soit exécuté,
> quel que soit leur emplacement dans le code. La portée d'une variable déclarée
> avec var est le contexte d'exécution courant, c'est-à-dire : la fonction qui
> contient la déclaration ou le contexte global si la variable est déclarée en
> dehors de toute fonction.
>
> Si on affecte une valeur à une variable qui n'a pas été déclarée (le mot-clé
> var n'a pas été utilisé), cela devient une variable globale (une propriété de
> l'objet global) lorsque l'affectation est exécutée.

[MDN - var](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Statements/var)

Si nous modifions le code pour ne plus utiliser `var` :

```js
z = 1;

function x() {
  y = 1;
  z = 2;
}

x();

console.log(y); // 1
console.log(z); // 2, car z est une variable globale
```

Autre détail, je n'ai pas réussi à trouver une source fiable pour dire que `var` peut être déclaré plusieurs fois [Re-Declaring JavaScript Variables](https://www.w3schools.com/js/js_variables.asp). Apparemment dans le premier exemple de code, `var z = 2;` ne redéclare pas `z` mais crée une nouvelle variable `z` dans le scope de la fonction `x`. C'est pour cela que `console.log(z);` affiche `1` et non `2`.

Est-ce que cela répond à votre question @Istomine ?

**Reply by Istomine** [reply]
Bonjour,

oui, ça répond à ma question. 

Merci
